src/goto-instrument/contracts/contracts.cpp

```diff
@@ -631,27 +631,44 @@ void code_contractst::replace_history_parameter(
   }
 }
 
-std::pair<goto_programt, goto_programt>
-code_contractst::create_ensures_instruction(
-  codet &expression,
-  source_locationt location,
-  const irep_idt &mode)
+void code_contractst::generate_history_variables_initialization(
+  exprt &clause,
+  const irep_idt &mode,
+  goto_programt &program)
 {
   std::map<exprt, exprt> parameter2history;
   goto_programt history;
-
   // Find and replace "old" expression in the "expression" variable
   replace_history_parameter(
-    expression, parameter2history, location, mode, history, ID_old);
-
-  // Create instructions corresponding to the ensures clause
-  goto_programt ensures_program;
-  converter.goto_convert(expression, ensures_program, mode);
+    clause, parameter2history, clause.source_location(), mode, history, ID_old);
+  // Add all the history variable initialization instructions
+  program.destructive_append(history);
+}
 
-  // return a pair containing:
-  // 1. instructions corresponding to the ensures clause
-  // 2. instructions related to initializing the history variables
-  return std::make_pair(std::move(ensures_program), std::move(history));
+void generate_contract_constraints(
+  code_contractst &contract,
```

Comment:
how about passing the converter directly instead of the whole `code_contractst` ?

Reply:
What about `add_quantified_variable`?

Reply:
my bad I missed that in the first pass